modules/dreamview/backend/perception_camera_updater/perception_camera_updater.cc

```diff
@@ -0,0 +1,221 @@
+/******************************************************************************
+ * Copyright 2019 The Apollo Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#include "modules/dreamview/backend/perception_camera_updater/perception_camera_updater.h"
+
+#include <float.h>
+#include <map>
+#include <string>
+#include "cyber/common/file.h"
+#include "modules/common/adapters/adapter_gflags.h"
+#include "opencv2/opencv.hpp"
+
+namespace apollo {
+namespace dreamview {
+
+using apollo::drivers::CompressedImage;
+using apollo::localization::LocalizationEstimate;
+using apollo::localization::Pose;
+using apollo::transform::TransformStampeds;
+using apollo::transform::Transform;
+
+namespace {
+  void ConvertMatrixToArray(
+      const Eigen::Matrix4d &matrix, std::vector<double> *array) {
+    const double* pointer = matrix.data();
+    for (int i = 0; i < matrix.size(); ++i) {
+      array->push_back(pointer[i]);
+    }
+  }
+}  // namespace
+
+PerceptionCameraUpdater::PerceptionCameraUpdater(WebSocketHandler* websocket)
+    : websocket_(websocket),
+      node_(cyber::CreateNode("perception_camera_updater")) {
+  InitReaders();
+}
+
+void PerceptionCameraUpdater::Start() {
+  enabled_ = true;
+}
+
+void PerceptionCameraUpdater::Stop() {
+  if (enabled_) {
+    localization_queue_.clear();
+    image_buffer_.clear();
+    tf_static_.clear();
+    current_image_timestamp_ = 0.0;
+  }
+  enabled_ = false;
+}
+
+void PerceptionCameraUpdater::GetImageLocalization(
+    std::vector<double> *localization) {
+  double timestamp_diff = DBL_MAX;
+  auto iter = localization_queue_.rbegin();
+  Pose image_pos;
+  for (; iter != localization_queue_.rend(); ++iter) {
+    const double tmp_diff = (*iter)->measurement_time()
+        - current_image_timestamp_;
+    if (tmp_diff > 0) {
+      timestamp_diff = tmp_diff;
+      image_pos = (*iter)->pose();
+    } else {
+      if (std::abs(tmp_diff) < timestamp_diff) {
+        image_pos = (*iter)->pose();
+      }
+      break;
+    }
+  }
+  // clean useless localization
+  auto iter_begin = localization_queue_.begin();
+  while ((*iter_begin) != (*iter)) {
+    iter_begin = localization_queue_.erase(iter_begin);
+  }
+
+  Eigen::Matrix4d localization_matrix;
```

Comment:
Extract this part as a utility function since static transform has similar codes